lib/spack/llnl/util/filesystem.py

```diff
@@ -265,25 +267,112 @@ def unset_executable_mode(path):
     os.chmod(path, mode)
 
 
+def copy(src, dest):
+    """Copies the file *src* to the file or directory *dest*.
+
+    If *dest* specifies a directory, the file will be copied into *dest*
+    using the base filename from *src*.
+
+    Parameters:
+        src (str): the file to copy
+        dest (str): the destination file or directory
+    """
+    tty.debug('Copying {0} to {1}'.format(src, dest))
+
+    shutil.copy(src, dest)
+
+
 def install(src, dest):
-    """Manually install a file to a particular location."""
-    tty.debug("Installing %s to %s" % (src, dest))
+    """Installs the file *src* to the file or directory *dest*.
+
+    If *dest* specifies a directory, the file will be installed into *dest*
+    using the base filename from *src*.
+
+    Works similarly to :py:func:`copy` with the addition of setting proper
+    permissions on the installed file.
+
+    Parameters:
+        src (str): the file to install
+        dest (str): the destination file or directory
+    """
+    tty.debug('Installing {0} to {1}'.format(src, dest))
 
     # Expand dest to its eventual full path if it is a directory.
     if os.path.isdir(dest):
         dest = join_path(dest, os.path.basename(src))
 
     shutil.copy(src, dest)
+
     set_install_permissions(dest)
     copy_mode(src, dest)
 
 
-def install_tree(src, dest, **kwargs):
-    """Manually install a directory tree to a particular location."""
-    tty.debug("Installing %s to %s" % (src, dest))
-    shutil.copytree(src, dest, **kwargs)
+def copy_tree(src, dest, symlinks=True):
+    """Recursively copy an entire directory tree rooted at *src*.
+
+    If the destination directory *dest* does not already exist, it will
+    be created as well as missing parent directories.
+
+    If *symlinks* is true, symbolic links in the source tree are represented
+    as symbolic links in the new tree and the metadata of the original links
+    will be copied as far as the platform allows; if false, the contents and
+    metadata of the linked files are copied to the new tree.
+
+    Parameters:
+        src (str): the directory to copy
+        dest (str): the destination directory
+        symlinks (bool): whether or not to preserve symlinks
+    """
+    tty.debug('Copying {0} to {1}'.format(src, dest))
+
+    mkdirp(dest)
+
+    for s, d in traverse_tree(src, dest, order='pre', follow_nonexisting=True):
+        if symlinks and os.path.islink(s):
+            # Note that this won't rewrite absolute links into the old
+            # root to point at the new root. Should we handle that case?
+            target = os.readlink(s)
+            os.symlink(os.path.abspath(target), d)
+        elif os.path.isdir(s):
+            mkdirp(d)
+        else:
+            shutil.copyfile(s, d)
+
+
+def install_tree(src, dest, symlinks=True):
```

Comment:
i would make these docs reference `copy_tree` and describe the differences, rather than repeating.

Reply:
I would also just implement this function by adding `permissions=False` argument to `copy_tree` and just calling `copy_tree` with it set to `True`, rather than copying the logic.  DRY, so that these functions don't diverge in weird ways in the future.